src/program.cpp

```diff
@@ -167,13 +168,37 @@ target_assignments program::get_target_assignments(const std::vector<target>& ta
     target_assignments p;
 
     const auto* mod = get_main_module();
-    for(auto it : iterator_for(*mod))
+    std::vector<supported_segments> target_subgraphs;
+    target_subgraphs.reserve(targets.size());
+    std::transform(targets.begin(),
+                   targets.end(),
+                   std::back_inserter(target_subgraphs),
+                   [mod, m](const auto& t) { return t.find_supported(mod, m); });
+
+    for(const auto ins : iterator_for(*mod))
     {
-        auto t = std::max_element(
-            targets.begin(), targets.end(), [it, m](const target& lhs, const target& rhs) {
-                return lhs.is_supported(it, m) < rhs.is_supported(it, m);
-            });
-        p.add_assignment(it, t->name());
+        if(p.has(ins))
+        {
+            continue;
+        }
+
+        for(auto i = 0U; i < target_subgraphs.size(); ++i)
+        {
+            const auto& subgraph = target_subgraphs[i];
+            const auto& target   = targets[i];
+            for(const auto& segment : subgraph)
+            {
+                const auto& instructions = segment.instructions;
+                if(instructions.find(ins) == instructions.end())
+                {
+                    continue;
+                }
+                for(const auto instr : instructions)
+                {
+                    p.add_assignment(instr, target.name());
+                }
```

Comment:
This should use `std::transform`:

```cpp
std::transform(instructions.begin(), instructions.end(), std::inserter(p, p.end()), [&](auto instr) {
    return std::make_pair(instr, target.name());
});
```

You need to add an `insert` method to the target assignments class.